File: peras-hs/src/Peras/Block.hs
{-# LANGUAGE DeriveGeneric #-}

module Peras.Block where

import Peras.Crypto (Hash(MkHash), Hashable, LeadershipProof, Signature(bytesS), VerificationKey)
import Peras.Numbering (RoundNumber, SlotNumber)


import GHC.Generics (Generic)
import Peras.Crypto (Hash (..), Hashable (..))
import Prelude hiding (round)

-- Use `Integer` for compatibility with `MAlonzo`.
type PartyId = Integer

data Party = MkParty{pid :: PartyId, pkey :: VerificationKey}
               deriving (Generic)

instance Eq Party where
    x == y = pid x == pid y && pkey x == pkey y

type Tx = ()

data Block = MkBlock{slotNumber :: SlotNumber,
                     creatorId :: PartyId, parentBlock :: Hash Block,
                     certificate :: Maybe Certificate,
                     leadershipProof :: LeadershipProof, signature :: Signature,
                     bodyHash :: Hash [Tx]}
               deriving (Generic)

data BlockBody = MkBlockBody{blockHash :: Hash [Tx],
                             payload :: [Tx]}
                   deriving (Generic)

data Certificate = MkCertificate{round :: RoundNumber,
                                 blockRef :: Hash Block}
                     deriving (Generic)

instance Eq Certificate where
    x == y = round x == round y && blockRef x == blockRef y

instance Eq Block where
    x == y
      = slotNumber x == slotNumber y &&
          creatorId x == creatorId y &&
            parentBlock x == parentBlock y &&
              leadershipProof x == leadershipProof y &&
                certificate x == certificate y &&
                  signature x == signature y && bodyHash x == bodyHash y

instance Eq BlockBody where
    x == y = blockHash x == blockHash y && payload x == payload y

instance Hashable Block where
    hash = \ b -> MkHash (bytesS (signature b))

File: peras-rust/src/block.rs
use crate::crypto::{Hash, LeadershipProof, Signature};
use serde::{Deserialize, Serialize};

pub type Slot = u64;

pub type PartyId = u64;

#[derive(Eq, Clone, PartialEq, Debug, Deserialize, Serialize, Hash)]
#[serde(tag = "tag", rename_all = "camelCase")]
pub struct Party {
    pub id: PartyId,
    #[serde(with = "hex::serde")]
    pub vkey: [u8; 8],
}

#[derive(Eq, Clone, PartialEq, Debug, Deserialize, Serialize, Hash)]
#[serde(tag = "tag", rename_all = "camelCase")]
pub struct Tx {}

#[derive(Eq, Clone, PartialEq, Debug, Deserialize, Serialize, Hash)]
#[serde(tag = "tag", rename_all = "camelCase")]
pub struct Block {
    pub slot_number: Slot,
    pub creator_id: PartyId,
    pub parent_block: Hash,
    pub certificate: Option<Certificate>,
    pub leadership_proof: LeadershipProof,
    pub signature: Signature,
    pub body_hash: Hash,
}

impl Block {
    pub fn genesis_hash() -> Hash {
        Hash {
            hash: [0, 0, 0, 0, 0, 0, 0, 0],
        }
    }
}

#[derive(Eq, Clone, PartialEq, Debug, Deserialize, Serialize, Hash)]
#[serde(tag = "tag", rename_all = "camelCase")]
pub struct BlockBody {
    pub block_hash: Hash,
    pub payload: Vec<Tx>,
}

#[derive(Eq, Clone, PartialEq, Debug, Deserialize, Serialize, Hash)]
#[serde(tag = "tag", rename_all = "camelCase")]
#[allow(non_snake_case)]
pub struct Certificate {
    pub voting_round_number: u64,
    pub blockRef: Hash,
}
